highlights(jsx): remove TSConstant from ecma and jsx

antonk52 opened the pull request:
closes #1547 

This PR removes `TSConstant` from `ecma` and `jsx` since it was based on casing and in the ecma family languages the casing does not affect the runtime. I have also grouped the queries using [alternations](https://tree-sitter.github.io/tree-sitter/using-parsers#alternations). This change does not affect the documentation.

The result can be seen below

![Screen Shot 2021-07-11 at 23 47 23](https://user-images.githubusercontent.com/5817809/125209601-9fe38700-e2a2-11eb-959e-d7dfae9c0276.png)

The one thing I could not figure out is the recursive queries. For example when the jsx element name can contain the nested identifier which incase can be another nested identifier like in the example  above called `member_expr_3`

antonk52 commented:
Hi, @kyazdani42 just a friendly ping

Any chance this PR can get a review? 😅

kyazdani42 commented:
Hi, i'm not really agreeing about this change. This breaks jsx/tsx highlighting quite a lot and in ecma family languages, there are nomenclature rules that are accepted by the community as common pattern. Highlighting jsx classes as tags makes it hard to know when these are react internals or user/libraries components.
If you feel strongly about this change, i would advise you to override theses queries in your own runtime :)

antonk52 commented:
Thank you for your comment. This PR was not to prove a point but to fix a potentially misleading highlighting. 

I understand your point regarding jsx components.

In the example I shared `TSConstant` is being applied to any identifier that starts with a capital letter. Besides the jsx example there is an example with a `class` and depending on a casing different highlighting groups get applied. This does not seem correct since it is statically analysable that Upper_Case is not a constant but a class identifier though can be reassign etc. Can we consider changing the logic behind highlighting identifiers based on their casing?

Anyway hope this helps 🙂

kyazdani42 commented:
Hi @antonk52 sorry for the late answer,
i guess your update would be appropriate enough. Although the current queries already make sense to me, because pascal case jsx is basically non react land components which, at first, where objects constructors, then classes (which are pascal cased in ecma languages). So highlighting as constructor makes sense, although today with function components i'm not so sure.
So go for it, if it keeps backward compatibility, i wouldn't be against it :)

antonk52 commented:
@kyazdani42 Thank you for your message

A summary of what I have done:
- updated my PR to preserve the constructor highlighting group and git history
- updated the PR name to reflect current changes
- introduced a new highlighting group `TSTagCustom` which links to `TSConstructor` by default to preserve backward compatibility for all themes that are already using it

A demo of what we have now
```viml
highlight TSTagCustom guifg=pink
highlight TSConstructor guifg=red
```
<img width="499" alt="Screen Shot 2021-08-08 at 21 25 26" src="https://user-images.githubusercontent.com/5817809/128645454-99169065-1650-465e-8207-35a9c63bd57d.png">


There are 2 difficulties I came across, any tips are welcome 🙂
1. I could not figure out how to use predicate within alternations
2. there are some PR checks failing, I've tried running them locally on master and they failed also. Not sure if my changes are related

Also noticed when speaking of jsx people usually refer to react. We should keep in mind that react is not the only lib/framework that uses jsx, svelte and angular also depend on it. Therefore perhaps we should avoid assumptions about assigning highlight groups to node names based on the implementation of a certain lib/framework. Though this is something that definitely needs a good triage first.

Hope this helps.

vigoux commented:
I am not sure we want to add yet another JSX specific highlight group.
My issue is that we keep adding more a more very specific highlight group, mainly only for the JSX/TSX/HTML trio, which is knid of cheating, while most other languages just try to use a limited amount of highlight groups.

We are not trying to provide full 100% accurate highlights here, but rather sane defaults, and for me these kind of changes (even if they I great, don't get me wrong), kind of fall in the "one new highlight group for 1 language, and 2 patterns in the query" pit.

I am not against the change, so I'll let others decide whether this change should be included or not, I am just raising a concern I have.

kyazdani42 commented:
I agree with @vigoux that adding more specific things just for the js family is kind of annoying. I'd like the input of @steelsojka on that too.

antonk52 commented:
This is a good point @vigoux, you can see in the [initial implementation](https://github.com/antonk52/nvim-treesitter/commit/518e2595a0dee308d935393cf8874aa79d6d0ab6) there were no new groups and some queries were removed. However, kyazdani42 noted that even though it fixes the issue it may break some themes that people are using already. This is why the current implementation introduces a new highlighting group with full backward compatibility.

If anyone can come up with a different solution to be able to differentiate custom components from non jsx TSConstant & TSConstructor groups I'd love consider it in this PR 🙂

amaanq commented:
Very stale and outdated, @antonk52 does this still interest you? I do agree though adding `@tag.custom` (no longer `TSTagCustom`) just for jsx is a bit awkward and too use-case specific

antonk52 commented:
@amaanq I still think this is a valid case. Given that `@tag.custom` inherits from `@tag` so it won't be a breaking change(to address my previous comment).

I am happy to update the branch if maintainers would be happy to merge it.

amaanq commented:
> @amaanq I still think this is a valid case. Given that `@tag.custom` inherits from `@tag` so it won't be a breaking change(to address my previous comment).
> 
> I am happy to update the branch if maintainers would be happy to merge it.

I'm ok with it, do update it if it's not a bother for you :)

ObserverOfTime commented:
I'd rather keep them as `@constructor` than introduce another highlight group.

clason commented:
Most of these changes are obsolete with the move of highlighting to Neovim core. If the query changes still make sense, please open a new PR from latest `master`.
